src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectStateChangeDetector.cs

```diff
@@ -229,7 +229,7 @@ protected virtual void InitializeSolution(Solution solution)
 
         private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
         {
-            if (args.Kind == ProjectChangeKind.ProjectAdded)
+            if (args.Kind == ProjectChangeKind.ProjectAdded || args.Kind == ProjectChangeKind.DocumentRemoved)
```

Comment:
Few questions:
- Why don't we need to check for DocumentAdded here? I imagine it'd have the same issue
- On a similar note, should we be having some similar check for the corresponding C# for the Razor file being added/removed in the workspace change kinds above: https://github.com/dotnet/aspnetcore-tooling/pull/3033/files#diff-66cb419a01c770c8a231f9de8074ee929416b8d912f4ba4887bbee6b6890ba40R110-R112
- Also similarly why do we "only" care about project added here? Might require some git blame history diving.

Reply:
TBF I jumped in here, proved that the behavior was still in the broken state from the issue, made a guess about what edit would fix it, and it did. It's entirely possible that this is the "wrong" fix.

1. Because DocumentAdded already works (I am unsure why).
2. I could do that, but what problems might be caused by NOT doing so and how might I test them?
3. That particular line [came from here](https://github.com/dotnet/aspnetcore-tooling/commit/a5621641f839dfa8e43c8d3d6b01552bf308a1b9) when `ProjectManager_Changed` was added. TBF I don't fully understand why this change was made, do you remember any context. If not I can go deeper on this.

It seems like in general I'm missing a lot of context about these StateChangeDetectors, so if we wanted to do a brain dump some time that might make things easier.

Reply:
I now know the "why" behind #1, when I added the component through the VS UI it ALSO opened the new file and changed focus to it, it was the DocumentOpened that was triggering the re-evaluation. So when I added a file out of the explorer it behaved like DocumentRemoved currently does, so we DO need to add DocumentAdded.

Reply:
Really appreciate you digging here to better understand this. It's such a key part of our system that fully understanding it is essential.

From your explanation it also sounds like adding a document via the windows explorer (basically outside of VS) wouldn't actually result in VS picking it up and reflecting appropriate changes too, so solid change!

Reply:
For #2 The scenario I could think of was: There's a static function in an @code block in Dingos.razor, you add a reference to it in FetchData.razor. Remove Dingos.razor, does a diagnostic show up? Re-add Dingos.razor, does the diagnostic disappear? It already behaved as desired in both cases,  so I think we're good there.

#3: I can't say for sure because the comments and PR history aren't very loud on the subject, but it seems like "only" caring about ProjectAdded is mostly just because we add things to this kind of Call-backy StateManagers in a very cautious "as needed" basis and it hadn't been called for before now.